src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java

```diff
@@ -0,0 +1,221 @@
+      }
+    });
+  }
+
+  abstract String getId(Transaction transaction);
+
+  abstract Criterion getCriterion(String value);
+
+  private Future<Tx<List<Transaction>>> createPermanentTransactions(Tx<List<Transaction>> tx) {
+    Promise<Tx<List<Transaction>>> promise = Promise.promise();
+    List<Transaction> transactions = tx.getEntity();
+    tx.getPgClient()
+      .saveBatch(tx.getConnection(), TRANSACTION_TABLE, transactions, reply -> {
+        if (reply.failed()) {
+          handleFailure(promise, reply);
+        } else {
+          promise.complete(tx);
+        }
+      });
+    return promise.future();
+  }
+
+  /**
+   * Accumulate transactions in a temporary table until expected number of transactions are present, then apply them all at once,
+   * updating all the required tables together in a database transaction.
+   * 
+   * @param transaction         processed transaction
+   * @param processTransactions The function responsible for the logic of saving transactions from the temporary table to the
+   *                            permanent one.
+   * @return completed future
+   */
+  Future<Void> processAllOrNothing(Transaction transaction,
+      Function<Tx<List<Transaction>>, Future<Tx<List<Transaction>>>> processTransactions) {
+    return createTempTransaction(transaction).compose(this::getSummary)
+      .compose(this::getTempTransactions)
+      .compose(transactions -> {
+        Promise<Void> promise = Promise.promise();
+        try {
+          if (isLastRecord) {
+            promise = moveFromTempToPermanentTable(processTransactions, transactions);
+          } else {
+            promise.complete();
+          }
+        } catch (Exception e) {
+          promise.fail(new HttpStatusException(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()));
+        }
+        return promise.future();
+      });
+  }
+
+  private Promise<Void> moveFromTempToPermanentTable(
+      Function<Tx<List<Transaction>>, Future<Tx<List<Transaction>>>> processTransactions, List<Transaction> transactions) {
+
+    Promise<Void> promise = Promise.promise();
+    Tx<List<Transaction>> tx = new Tx<>(transactions, getPostgresClient());
+
+    startTx(tx).compose(processTransactions)
+      .compose(this::deleteTempTransactions)
+      .compose(HelperUtils::endTx)
+      .setHandler(result -> {
+        if (result.failed()) {
+          HttpStatusException cause = (HttpStatusException) result.cause();
+          log.error("Transactions {} or associated data failed to be processed", cause, tx.getEntity());
+
+          rollbackTransaction(tx).setHandler(res -> promise.fail(cause));
+        } else {
+          log.info("Transactions {} and associated data were successfully processed", tx.getEntity());
+          promise.complete();
+        }
+      });
+
+    return promise;
+  }
+
+  private Future<Transaction> createTempTransaction(Transaction transaction) {
```

Comment:
I see , we are not storing the purchaseOrderId in the table in a separate column, which is fine. just make sure to update the WIKI

Reply:
Actually we are, but under a different name - `encumbrance_sourcePurchaseOrderId`. It is generated by RMB from foreign key defenition.